include/circt/Dialect/RTL/Types.td

```diff
@@ -0,0 +1,96 @@
+//===- Types.td - RTL data type definitions ----------------*- tablegen -*-===//
+//
+// Basic data types for the RTL dialect.
+//
+//===----------------------------------------------------------------------===//
+
+// Base class for other typedefs. Provides dialact-specific defaults
+class RTLType<string name> : TypeDef<RTLDialect, name> { }
+
+// A type with an array of `FieldInfo`s. Essentially, a parent class for the
+// struct and union types.
+class FieldInfo_Type<string name> : RTLType<name> {
+  let parameters = (
+    ins
+    ArrayRefOfSelfAllocationParameter<"::circt::rtl::FieldInfo", "">: $fields
+  );
+
+  let printer = [{
+    printer << "struct" << "<";
+    for (size_t i=0; i<getImpl()->fields.size(); i++) {
+      const auto& field = getImpl()->fields[i];
+      printer << "{" << field.name << "," << field.type << "}";
+      if (i < getImpl()->fields.size() - 1)
+        printer << ",";
+    }
+    printer << ">";
+  }];
+
+  let parser = [{
+    SmallVector<FieldInfo, 4> parameters;
+    if (parser.parseLess()) return ::mlir::Type();
+    while (mlir::succeeded(parser.parseOptionalLBrace())) {
+      StringRef name;
+      if (parser.parseKeyword(&name)) return ::mlir::Type();
+      if (parser.parseComma()) return ::mlir::Type();
+      ::mlir::Type type;
+      if (parser.parseType(type)) return ::mlir::Type();
+      if (parser.parseRBrace()) return ::mlir::Type();
+      parameters.push_back(FieldInfo(name, type));
+      if (parser.parseOptionalComma()) break;
+    }
+    if (parser.parseGreater()) return ::mlir::Type();
+    return get(ctxt, parameters);
+  }];
+}
+
+// A struct type. Inherits the parser/printer from FieldInfo_Type.
+def StructType : FieldInfo_Type<"Struct"> {
+  let summary = "A typical structure type";
+  let description = [{
+    A C-like or SystemVerilog-like struct. Missing alignment specification.
+    We will add this once we figure out what exactly this means in hardware.
+    (It may be obvious, we just don't want to make any assumptions.)
+  }];
+  let mnemonic = "struct";
+}
+
+// A union type. Inherits the parser/printer from FieldInfo_Type.
+def UnionType : FieldInfo_Type<"Union"> {
```

Comment:
Please split out the addition of types to the RTL dialect to its own path and discussion.  I think we should start with one of them (e.g. structs or arrays) to anchor the conversation and then tackle the rest in a follow-on.